File: BuddyGrammarKit/Sources/BuddyGrammarKit/AutomaticSuggestionReplacement.swift
import Foundation

/// Immutable metadata captured when a replacement suggestion is rendered.
///
/// Production adapters validate the exact bounded editor observation before
/// deleting anything. This prevents a stale strip candidate from replacing an
/// unrelated suffix that merely happens to have the same character count.
public struct AutomaticSuggestionReplacement: Equatable, Sendable {
    public let originalText: String
    public let replacementText: String
    public let boundary: String
    public let precedingContext: String
    public let source: AutomaticCorrectionSource

    public init?(
        originalText: String,
        replacementText: String,
        boundary: String,
        precedingContext: String,
        source: AutomaticCorrectionSource
    ) {
        guard !originalText.isEmpty,
              !replacementText.isEmpty,
              originalText != replacementText else {
            return nil
        }
        self.originalText = originalText
        self.replacementText = replacementText
        self.boundary = boundary
        self.precedingContext = precedingContext
        self.source = source
    }

    public var expectedContextBeforeInput: String {
        precedingContext + originalText
    }

    public var insertion: String {
        replacementText + boundary
    }

    public func matches(
        contextBeforeInput: String?,
        deleteCount: Int,
        insertion: String
    ) -> Bool {
        contextBeforeInput == expectedContextBeforeInput
            && deleteCount == originalText.count
            && insertion == self.insertion
    }
}

File: android/app/src/main/java/com/francescooddo/buddygrammar/core/AutomaticSuggestionReplacement.kt
package com.francescooddo.buddygrammar.core

enum class AutomaticSuggestionSource(val receiptValue: String) {
    TAP_LATTICE("tapLattice"),
    SPELLING("spelling"),
    SHORTCUT("shortcut"),
    SWIPE("swipe"),
}

/**
 * Immutable metadata captured with a rendered replacement suggestion.
 *
 * Requiring the exact bounded editor observation before deletion prevents a
 * stale strip candidate from replacing an unrelated suffix with the same
 * UTF-16 length.
 */
@ConsistentCopyVisibility
data class AutomaticSuggestionReplacement private constructor(
    val originalText: String,
    val replacementText: String,
    val boundaryText: String,
    val precedingContext: String,
    val source: AutomaticSuggestionSource,
    val renderedFieldEpoch: Long? = null,
    val renderedFieldIdentifier: String? = null,
    val renderedLanguageTag: String? = null,
) {
    val expectedContextBeforeCursor: String get() = precedingContext + originalText
    val insertion: String get() = replacementText + boundaryText

    fun matches(
        contextBeforeCursor: String,
        replaceBeforeCursor: Int,
        insertion: String,
    ): Boolean =
        contextBeforeCursor == expectedContextBeforeCursor &&
            replaceBeforeCursor == originalText.length &&
            insertion == this.insertion

    /** Adds the editor ownership captured when this candidate enters the UI. */
    fun ownedBy(
        fieldEpoch: Long,
        fieldIdentifier: String,
        languageTag: String,
    ): AutomaticSuggestionReplacement {
        require(fieldIdentifier.isNotBlank()) { "fieldIdentifier must not be blank" }
        require(languageTag.isNotBlank()) { "languageTag must not be blank" }
        return copy(
            renderedFieldEpoch = fieldEpoch,
            renderedFieldIdentifier = fieldIdentifier,
            renderedLanguageTag = languageTag,
        )
    }

    fun isOwnedBy(
        fieldEpoch: Long,
        fieldIdentifier: String,
        languageTag: String,
    ): Boolean = renderedFieldEpoch == fieldEpoch &&
        renderedFieldIdentifier == fieldIdentifier &&
        renderedLanguageTag == languageTag

    companion object {
        fun create(
            originalText: String,
            replacementText: String,
            boundaryText: String,
            precedingContext: String,
            source: AutomaticSuggestionSource,
        ): AutomaticSuggestionReplacement? {
            if (
                originalText.isEmpty() || replacementText.isEmpty() ||
                originalText == replacementText
            ) return null
            return AutomaticSuggestionReplacement(
                originalText = originalText,
                replacementText = replacementText,
                boundaryText = boundaryText,
                precedingContext = precedingContext,
                source = source,
            )
        }
    }
}
